File: java/src/eti/Project2/Organisms/Guarana.java
package eti.Project2.Organisms;

import eti.Project2.GUI;
import eti.Project2.MapPosition;
import eti.Project2.World;

/**
 * Created By IDEA
 * Author: s188026 Yifei Liu
 * Date: 2021/5/30 14:06
 * Description: Guarana     '#':
 *      Strength of the animal which ate guarana is permanently increased by 3.
 */

public class Guarana extends Plant{
    //constructor
    public Guarana(char n, int s, int i, int X, int Y, World w) {
        super(n, s, i, X, Y, w);
    }

    public Guarana(Guarana g) {
        super(g);
    }

    public Guarana (int X, int Y, World world) {
        setName('#');
        setStrength(0);
        setInitiative(0);
        setPosition(new MapPosition(X, Y));
        setWorld(world);
    }

    @Override
    public void Collision(Organism attacker)
    {
        //default collision
        if (attacker.getStrength() >= this.getStrength())
        {
            getWorld().organisms[getPosition().getY()][getPosition().getX()] = attacker;

            /**std::cout + "'" + getName() + "'(" + getPosition().getX() + "," + getPosition().getY() + ") increases 3 strength of attacker who eats it'" + (Grid(attacker.getPosition().getX(), attacker.getPosition().getY())).getName() + "'(" + attacker.getPosition().getX() + "," + attacker.getPosition().getY() + ")" + std::endl;*/
            GUI.getInformation().append("'" + getName() + "'(" + getPosition().getX() + "," + getPosition().getY() + ") increases 3 strength of attacker who eats it'" + attacker.getName() + "'(" + attacker.getPosition().getX() + "," + attacker.getPosition().getY() + ")\n");
            
            attacker.setPosition(this.getPosition());
            attacker.setStrength(attacker.getStrength() + 3);
        }
    }
}

File: cpp/Guarana.cpp
#include "Guarana.h"
#include <iostream>

Guarana::Guarana(int X, int Y, World* world) :Plant(GUARANA, 0, 0, X, Y, world) {};

void Guarana::Collision(Organism*& attacker)
{
	//default collision
	if (attacker->GetStrength() >= this->GetStrength())
	{
		Grid(GetPosition().X, GetPosition().Y) = attacker;

		//if = -1, it is nullorgan or baby of two same animal(here is !=)
		/*if (attacker->GetPosition().X != -1)
			Grid(attacker->GetPosition().X, attacker->GetPosition().Y) = new nullGrid(attacker->GetPosition());*/

		//if (attacker->GetPosition().X == -1)
		//	this->SetAge(0);				//same strength, attacker win, so baby win(even if there is something let this be stronger, attacker will not pass code to here)

		std::cout << "'" << GetName() << "'(" << GetPosition().X << "," << GetPosition().Y << ") increases 3 strength of attacker who eats it'" << (Grid(attacker->GetPosition().X, attacker->GetPosition().Y))->GetName() << "'(" << attacker->GetPosition().X << "," << attacker->GetPosition().Y << ")" << std::endl;

		attacker->SetPosition(this->GetPosition());
		attacker->SetStrength(attacker->GetStrength() + 3);
		
		delete(this);

		//return false;					//attacker win return false
	}
	else
	{
		//if = -1, it is nullorgan or baby of two same animal
		/*if (attacker->GetPosition().X != -1)
			Grid(attacker->GetPosition().X, attacker->GetPosition().Y) = new nullGrid(attacker->GetPosition());*/

		delete(attacker);

		//return true;					//defender win return true
	}
}